src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -81,13 +88,29 @@ internal class SemanticTokensHelpers
             // We either calculate the tokens for the full document span, or the user 
             // can pass in a range from the full document if they wish.
             ranges ??= new[] { ProtocolConversions.TextSpanToRange(root.FullSpan, text) };
+            if (s_stopExperiment || s_random.Next(2) == 0)
+            {
+                var textSpans = new TextSpan[ranges.Length];
+                for (var i = 0; i < ranges.Length; i++)
+                {
+                    textSpans[i] = ProtocolConversions.RangeToTextSpan(ranges[i], text);
+                }
 
-            foreach (var range in ranges)
+                var stopwatch = Stopwatch.StartNew();
+                await GetClassifiedSpansForDocumentAsync(
+                    classifiedSpans, document, textSpans, options, cancellationToken).ConfigureAwait(false);
+                s_new.Add(stopwatch.ElapsedMilliseconds);
+            }
+            else
             {
+                var range = new LSP.Range { Start = ranges[0].Start, End = ranges[ranges.Length - 1].End };
                 var textSpan = ProtocolConversions.RangeToTextSpan(range, text);
+                var textSpans = new TextSpan[] { textSpan };
 
+                var stopwatch2 = Stopwatch.StartNew();
                 await GetClassifiedSpansForDocumentAsync(
-                    classifiedSpans, document, textSpan, options, cancellationToken).ConfigureAwait(false);
+                    classifiedSpans, document, textSpans, options, cancellationToken).ConfigureAwait(false);
+                s_old.Add(stopwatch2.ElapsedMilliseconds);
```

Comment:
note: this is not threadsafe, and may corrupt.  should do `lock (s_old) ...` (same with s_new).